### PR TITLE
[Homework] Make Shader API code independent to OpenGL code.

### PR DESCRIPTION
**Pull request by 3x3y3z3t**
At the end of one of the episodes The Cherno have given us a "homework" which is to seperate the OpenGL code from Shader class to make it platform independent. Since I haven't seen him touch that topic for a few episodes I think I should finish the homework and keep Shader in line with the other components.
Also this fix the problem mentioned in PR #72: added a missing semicolon in LayerStack.cpp

### CODE REVIEW

File: Hazel/src/Hazel/LayerStack.cpp [hunk 1]
```diff
@@ -28,18 +28,18 @@ namespace Hazel {
 
 	void LayerStack::PopLayer(Layer* layer)
 	{
-		auto it = std::find(m_Layers.begin(), m_Layers.end(), layer);
+		auto it = std::find(m_Layers.begin(), m_Layers.begin() + m_LayerInsertIndex, layer);
 		if (it != m_Layers.end())
 		{
-			layer->OnDetach()
+            layer->OnDetach();
```
Comment (LovelySanta) [inline Hazel/src/Hazel/LayerStack.cpp]:
```suggestion
		layer->OnDetach();
```

File: Hazel/src/Hazel/Renderer/Shader.h [hunk 2]
```diff
@@ -7,13 +7,10 @@ namespace Hazel {
 	class Shader
 	{
 	public:
-		Shader(const std::string& vertexSrc, const std::string& fragmentSrc);
-		~Shader();
+        virtual void Bind() const = 0;
```
Comment (LovelySanta) [inline Hazel/src/Hazel/Renderer/Shader.h]:
```suggestion
		virtual void Bind() const = 0;
```

File: Hazel/src/Hazel/Renderer/Shader.h [hunk 3]
```diff
@@ -7,13 +7,10 @@ namespace Hazel {
...
+		virtual void Unbind() const = 0;
 
-		void Bind() const;
-		void Unbind() const;
-	private:
-		uint32_t m_RendererID;
+        static Shader* Create(const std::string& vertexSrc, const std::string& fragmentSrc);
```
Comment (LovelySanta) [inline Hazel/src/Hazel/Renderer/Shader.h]:
```suggestion
		static Shader* Create(const std::string& vertexSrc, const std::string& fragmentSrc);
```

**Comment by 3x3y3z3t** [comment]
Sorry for the spaces - tabs mismatch.
I'm using 4 spaces indent in my local workspace, and everything should be formatted (indent) correctly. The mismatch may be because of me editing files directly on github web, which the editor is set to tabs by default, and I didn't noticed that.

**Comment by LovelySanta** [comment]
As mensioned in #52, premake could resolve this.

**Comment by LovelySanta** [comment]
Just a note to myself and maybe to cherno: A place on the github wiki; one page that list a possible solution for the homework assignment. So others can make a pull request (with homework flag) to propose a change, and a good answer can be picked to put up in the wiki for future references

**Comment by 3x3y3z3t** [comment]
As [The Cherno have implemented Shader abstraction](https://github.com/TheCherno/Hazel/commit/3c147806f0b253e4a288b64e65a81948d63f112e), this PR become obsolete and need to be closed.

**Pull request closed without merging**